File: src/spirite/base/pen/behaviors/PickBehavior.java
package spirite.base.pen.behaviors;

import spirite.base.graphics.RawImage;
import spirite.base.graphics.renderer.RenderEngine.RenderSettings;
import spirite.base.pen.Penner;
import spirite.base.util.MUtil;

public class PickBehavior extends StateBehavior {
	/**
	 * 
	 */
	private final Penner penner;
	final boolean leftClick;
	
	public PickBehavior( Penner penner, boolean leftClick) {
		super(penner);
		this.penner = penner;
		this.leftClick = leftClick;
	}
	
	private void pickColor() {
		// Get the composed image
		RenderSettings settings = new RenderSettings(
				this.penner.renderEngine.getDefaultRenderTarget(this.penner.workspace));
		RawImage img = this.penner.renderEngine.renderImage(settings);
		
		if( !MUtil.coordInImage(this.penner.x, this.penner.y, img))
			return;
		this.penner.paletteManager.setActiveColor( (leftClick)?0:1, img.getRGB(this.penner.x, this.penner.y));
	}

	@Override
	public void start() {
		pickColor();
	}
	@Override
	public void onMove() {
		pickColor();
	}
	@Override public void onTock() {}
}

File: src/main/kotlin/spirite/base/pen/behaviors/PickBehavior.kt
package spirite.base.pen.behaviors

import spirite.base.pen.Penner
import spirite.base.util.linear.Rect

class PickBehavior(penner: Penner, val leftClock: Boolean) : PennerBehavior(penner) {

    fun pick() {
        val img = penner.renderEngine.renderWorkspace(penner.workspace ?: return)
        if( Rect(img.width, img.height).contains(penner.x, penner.y))
            penner.paletteManager.activeBelt.setColor( if( leftClock) 0 else 1, img.getColor(penner.x, penner.y))
    }

    override fun onStart() {pick()}
    override fun onTock() {}
    override fun onMove() { pick()}
}